flatland/envs/agent_chains.py

```diff
@@ -49,66 +71,42 @@ def addAgent(self, iAg, rc1, rc2, xlabel=None):
         self.G.add_edge(rc1, rc2)
         self.Grev.add_edge(rc2, rc1)
 
-    def find_stops(self):
-        """ find all the stopped agents as a set of rc position nodes
-            A stopped agent is a self-loop on a cell node.
+    def find_stopped_agents(self) -> Set[Cell]:
         """
-
-        # get the (sparse) adjacency matrix
-        spAdj = nx.linalg.adjacency_matrix(self.G)
-
-        # the stopped agents appear as 1s on the diagonal
-        # the where turns this into a list of indices of the 1s
-        giStops = np.where(spAdj.diagonal())[0]
-
-        # convert the cell/node indices into the node rc values
-        lvAll = list(self.G.nodes())
-        # pick out the stops by their indices
-        lvStops = [lvAll[i] for i in giStops]
-        # make it into a set ready for a set intersection
-        svStops = set(lvStops)
-        return svStops
-
-    def find_stops2(self):
-        """ alternative method to find stopped agents, using a networkx call to find selfloop edges
+        Find stopped agents, using a networkx call to find self-loop nodes.
+        :return: set of stopped agents
         """
         svStops = {u for u, v in nx.classes.function.selfloop_edges(self.G)}
         return svStops
 
-    def find_stop_preds(self, svStops=None):
-        """ Find the predecessors to a list of stopped agents (ie the nodes / vertices)
-            Returns the set of predecessors.
-            Includes "chained" predecessors.
+    def find_stop_preds(self, svStops: Set[Cell]) -> Set[Cell]:
+        """ Find the predecessors to a list of stopped agents (ie the nodes / vertices). Includes "chained" predecessors.
+            :param svStops: list of voluntarily stopped agents
+            :return: the set of predecessors.
         """
 
-        if svStops is None:
-            svStops = self.find_stops2()
-
         # Get all the chains of agents - weakly connected components.
         # Weakly connected because it's a directed graph and you can traverse a chain of agents
         # in only one direction
+        # TODO why do we need weakly connected components at all? Just use reverse traversal of directed edges?
```

Comment:
@hagrid67 @aiAdrian why do we need weakly connected components at all? We could just reverse traverse the directed graph?

Reply:
Yes I think you are right once again!  We could just traverse the reverse graph, starting from the "swaps and stops" (ie immediate deadlocks, and voluntary / failed / slow agents).  There doesn't seem to be any need to break the graph down into weakly connected components first.  The traversal will naturally terminate as it reaches the "boundaries" of WCCs.

When I wrote the motioncheck I tried to look for algos which did clever stuff, assuming they would be faster than my own implementation.  Processing the WCCs separately seemed reasonable and my small-case performance checks did not expose the O(n^2) performance problems we seem to have.